### DISCUSSION TITLE
Question about the intended behavior of Accuracy with `average="none"` and `nan` results

### DISCUSSION THREAD

**Discussion by phiyodr** [Lightning-AI/torchmetrics#466] [Classification] [answered]
Hi,

I'm using your Accuracy metric without averaging and have difficulties with `nan` results, as I want to process the tensor further. 

Example with 4 binary labels (cat, dog, horse, cow) and 3 samples:

```python
from torchmetrics import Accuracy # version '0.5.0'
accs = Accuracy(average="none", num_classes=4)

probs  = torch.tensor([[.9, 0.1, 0.9, 0.1], 
                       [.8, 0.2, 0.9, 0.2], 
                       [.7, 0.1, 0.8 , 0.2]])

targets = torch.tensor([[1, 1, 0, 0], 
                        [1, 1, 0, 0], 
                        [1, 1, 0, 0]])

accs(probs, targets)
#> tensor([1., 0., 0., nan])
```

On your [Docs](https://torchmetrics.readthedocs.io/en/latest/references/modules.html?highlight=accuracy#id2) you write that this is intended: "_If 'none' and a given class doesn’t occur in the preds or target, the value for the class will be nan_."

But technically the last class has an Accuracy of `1.`. Statistics like `min` or `max` do not work on tensors with `nan`. My real model has some thousand labels and imputing `nan` with `1.` doesn't feel that good ;)

Internally you treat the last label with an Accuracy of `1.`:  

```python
acc = Accuracy(num_classes=4)
acc(probs, targets)
#> tensor(0.5000)
``` 

To some extend I understand your thoughts, but I want to ask if this is really the intended result or if an argument like `suppress_nan` is possible.
Thanks for your module!

Best 
Philipp

**Comment by SkafteNicki** [comment]
Hi @phiyodr,
Thanks for raising this issue. Looking at our code it may very well be that there is some kind of bug here. 
I am pretty sure the comment *"If 'none' and a given class doesn’t occur in the preds or target, the value for the class will be nan."* is only meant to apply to multi-class problems, and not multi-label problems as you are showing.

**Reply by phiyodr** [reply]
Thank you! I will make it an issue.

**Comment by Borda** [comment, marked as answer]
@phiyodr — as @SkafteNicki noted, the NaN behavior for absent classes was a known limitation. In v1.9.0, this is handled via the `zero_division` parameter:
```python
from torchmetrics.classification import MultilabelAccuracy

acc = MultilabelAccuracy(
    num_labels=4,
    average="none",
    zero_division=0.0,  # return 0 instead of NaN for absent classes
)

per_class = acc(preds, target)
print(per_class.min(), per_class.max())  # no NaN issues
```

`zero_division` is available on all classification metrics (F1, Precision, Recall, Accuracy, etc.):

- `zero_division=0.0` — treat absent classes as score 0 (conservative)
- `zero_division=1.0` — treat absent classes as score 1 (optimistic)
- Default (NaN) — flag absent classes explicitly
```python
from torchmetrics.classification import MultilabelF1Score

f1 = MultilabelF1Score(num_labels=4, average="none", zero_division=0.0)
per_label_f1 = f1(preds, target)  # no NaN
```

Docs: [MultilabelAccuracy](https://lightning.ai/docs/torchmetrics/stable/classification/accuracy.html)
